File: Components/DevExpressVCL/Demos/ExpressBars/Delphi/InPlaceEditorsDemo/InPlaceEditorsDemoComboBoxes.pas
unit InPlaceEditorsDemoComboBoxes;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, ExtCtrls, cxGraphics, cxGeometry, Math, InPlaceEditorsDemoFrameManager,
  StdCtrls;

type
  TfrmComboBoxes = class(TEditorDemoBaseFrame)
    PaintBox1: TPaintBox;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure PaintBox1Paint(Sender: TObject);
  private
    FPath: string;
    FFontName: string;
    FFontSize: Integer;
    FColor: TColor;
    FImage: TBitmap;
    FPerson: string;
  public
    procedure SetParameters(AColor: TColor; const AFontName, AFolder, APerson: string; AFontSize: Integer; AImage: TBitmap);
  end;

implementation

{$R *.dfm}

uses
  Types;

{ TfrmComboBoxes }

procedure TfrmComboBoxes.SetParameters(AColor: TColor; const AFontName, AFolder, APerson: string; AFontSize: Integer; AImage: TBitmap);
begin
  FColor := AColor;
  FFontName := AFontName;
  FFontSize := AFontSize;
  FPath := AFolder;
  FImage.Assign(AImage);
  FPerson := Format('Contact person: %s', [APerson]);
  PaintBox1.Invalidate;
end;

procedure TfrmComboBoxes.FormCreate(Sender: TObject);
begin
  FImage := TBitmap.Create;
end;

procedure TfrmComboBoxes.FormDestroy(Sender: TObject);
begin
  FImage.Free;
end;

procedure TfrmComboBoxes.PaintBox1Paint(Sender: TObject);
var
  ARect: TRect;
  AIndent: Integer;
begin
  if FImage <> nil then
  begin
    PaintBox1.Canvas.Brush.Color := FColor;
    PaintBox1.Canvas.FillRect(PaintBox1.ClientRect);
    PaintBox1.Canvas.Font.Name := FFontName;
    PaintBox1.Canvas.Font.Size := FFontSize;
    PaintBox1.Canvas.Brush.Style := bsClear;
    AIndent := Min((PaintBox1.Width - FImage.Width - PaintBox1.Canvas.TextWidth(FPerson) - 4) div 2,
      (PaintBox1.Width - FImage.Width - PaintBox1.Canvas.TextWidth(FPath) - 4) div 2);
    AIndent := Max(0, AIndent);
    ARect := cxRectBounds(AIndent, (PaintBox1.Height - FImage.Height) div 2, FImage.Width, FImage.Height);
    cxDrawImage(PaintBox1.Canvas.Handle, ARect, ARect, FImage, nil, -1, idmNormal);
    PaintBox1.Canvas.TextOut(ARect.Right + 4, (PaintBox1.Height - PaintBox1.Canvas.TextHeight(FPath)) div 2, FPath);
    PaintBox1.Canvas.TextOut(ARect.Right + 4, (PaintBox1.Height - PaintBox1.Canvas.TextHeight(FPerson)) div 2 + PaintBox1.Canvas.TextHeight(FPath), FPerson);
  end;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressBars/CBuilder/InPlaceEditorsDemo/InPlaceEditorsDemoComboBoxes.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop
#include <math.hpp>
#include "cxGraphics.hpp"
#include "cxGeometry.hpp"
#include "InPlaceEditorsDemoComboBoxes.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmComboBoxes *frmComboBoxes;
//---------------------------------------------------------------------------
__fastcall TfrmComboBoxes::TfrmComboBoxes(TComponent* Owner)
	: TEditorDemoBaseFrame(Owner)
{
  FImage = new Graphics::TBitmap;
}
//---------------------------------------------------------------------------
void __fastcall TfrmComboBoxes::SetParameters(TColor AColor, AnsiString AFontName, AnsiString AFolder,
	  int AFontSize, Graphics::TBitmap* AImage, String APerson)
{
  FColor = AColor;
  FFontName = AFontName;
  FFontSize = AFontSize;
  FPath = AFolder;
  FPerson = "Contact person: " + APerson;
  FImage->Assign(AImage);
  PaintBox1->Invalidate();
}

void __fastcall TfrmComboBoxes::FormDestroy(TObject *Sender)
{
  delete FImage;
}
//---------------------------------------------------------------------------

void __fastcall TfrmComboBoxes::PaintBox1Paint(TObject *Sender)
{
  if (FImage != 0)
  {
    PaintBox1->Canvas->Brush->Color = FColor;
    PaintBox1->Canvas->FillRect(PaintBox1->ClientRect);
    PaintBox1->Canvas->Font->Name = FFontName;
    PaintBox1->Canvas->Font->Size = FFontSize;
    PaintBox1->Canvas->Brush->Style = bsClear;
    int AIndent = Min((PaintBox1->Width - FImage->Width - PaintBox1->Canvas->TextWidth(FPerson) - 4) / 2,
      (PaintBox1->Width - FImage->Width - PaintBox1->Canvas->TextWidth(FPath) - 4) / 2);
    AIndent = Max(0, AIndent);
	TRect ARect = cxRectBounds(AIndent, (PaintBox1->Height - FImage->Height) / 2, FImage->Width, FImage->Height);
	cxPaintCanvas()->BeginPaint(PaintBox1->Canvas);
	cxDrawImage(cxPaintCanvas(), ARect, FImage, 0, -1, true);
	cxPaintCanvas()->EndPaint();
	PaintBox1->Canvas->TextOut(ARect.Right + 4, (PaintBox1->Height - PaintBox1->Canvas->TextHeight(FPath)) / 2, FPath);
    PaintBox1->Canvas->TextOut(ARect.Right + 4, (PaintBox1->Height - PaintBox1->Canvas->TextHeight(FPerson)) / 2 + PaintBox1->Canvas->TextHeight(FPath), FPerson);
  }
}
//---------------------------------------------------------------------------

void __fastcall TfrmComboBoxes::FormCreate(TObject *Sender)
{
//
}
//---------------------------------------------------------------------------

